node/src/store/sled_store.rs

```diff
@@ -0,0 +1,392 @@
+use std::convert::Infallible;
+use std::ops::Deref;
+use std::path::Path;
+
+use async_trait::async_trait;
+use celestia_types::hash::Hash;
+use celestia_types::ExtendedHeader;
+use sled::transaction::{ConflictableTransactionError, TransactionError};
+use sled::{Db, Error as SledError, Transactional, Tree};
+use tendermint_proto::Protobuf;
+use tracing::debug;
+
+use crate::store::Store;
+use crate::store::{Result, StoreError};
+
+const HEAD_HEIGHT_KEY: &[u8] = b"KEY.HEAD_HEIGHT";
+const HASH_TREE_ID: &[u8] = b"HASH";
+const HEIGHT_TO_HASH_TREE_ID: &[u8] = b"HEIGHT";
+
+#[derive(Debug)]
+pub struct SledStore {
+    db: Db,
+    headers: Tree,
+    height_to_hash: Tree,
+}
+
+impl SledStore {
+    pub fn new<P>(path: P) -> sled::Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let db = sled::open(path)?;
+        Self::init(db)
+    }
+
+    pub fn new_in_memory() -> sled::Result<Self> {
+        let db = sled::Config::default().temporary(true).open()?;
```

Comment:
this is interesting. So with such config sled will use the `/dev/shm`. I have mixed feelings about that behavior. It sounds cool but I'm not sure if the `/tmp` wouldn't be better match as we expect this to be of non-trivial size.
This on the other hand sounds like what @oblique would love the `InMemoryStore` to behave by still serializing the headers

Reply:
I think we shouldn't touch the default dehavior of sled, because it is [implemented in a cross-platform way](https://github.com/spacejam/sled/blob/005c023ca94d424d8e630125e4c21320ed160031/src/config.rs#L412-L417). `/tmp` is not really cross-platform in UNIX systems. For example in macOS it is cleaned periodically, and programs need to use `$TMPDIR` that points to a different directory to avoid that.

Reply:
yeah I didn't want to go with a raw `/tmp` but rather some `tempdir` crate anyway. I'm wondering if ram is what we want

Reply:
I've added `new_in_memory` mostly with tests in mind, but with platform discrepancies we should probably avoid that as Yiannis mentioned. Changed the API to suggested :+1: